File: Behavioral/Iterator/python/iterator.py
class Item:

    def __init__(self, name, val):
        self.name = name
        self.val = val

    def __str__(self):
        return "{} {}".format(self.name, self.val)

class ItemsIterator:
    def __init__(self, items):
        self.index = 0
        self.items = items

    def has_next(self):
        if self.index >= len(self.items):
            return False
        return True

    def next(self):
        item = self.items[self.index]
        self.index += 1
        return item

if __name__ == '__main__':
    item1 = Item('car', 30000)
    item2 = Item('house', 100000)
    item3 = Item('bike', 500)
    item4 = Item('tv', 800)
    items = []
    items.append(item1)
    items.append(item2)
    items.append(item3)
    items.append(item4)

    iterator = ItemsIterator(items)

    while iterator.has_next():
        item = iterator.next()
        print item

File: Behavioral/Iterator/Go/iterator.go
package main

import "fmt"

// Iterator interface defines methods for iterator pattern
type Iterator interface {
	// HasNext returns true if there are any items left to be iterated
	HasNext() bool

	// Next returns the next item in the collection
	Next() interface{}
}

// Collection defines interface for collection of items
type Collection interface {
	// Add adds item into the collection
	Add(item interface{})

	// Iterator returns iterator for the Collection
	Iterator() Iterator
}

// List defines list of items
type List struct {
	items []interface{}
}

// ListIterator holds the data needed to iterate a List
type ListIterator struct {
	index int
	items []interface{}
}

func (i *ListIterator) HasNext() bool {
	return i.index < len(i.items)
}

func (i *ListIterator) Next() interface{} {
	if !i.HasNext() {
		return nil
	}

	item := i.items[i.index]
	i.index++
	return item
}

func (c *List) Add(item interface{}) {
	c.items = append(c.items, item)
}

func (c *List) Iterator() Iterator {
	return &ListIterator{items: c.items}
}

func main() {
	list := &List{}
	list.Add(10)
	list.Add(3.14)
	list.Add("hello")

	itr := list.Iterator()
	for ; itr.HasNext(); {
		fmt.Println(itr.Next())
	}

	// prints:
	// 10
	// 3.14
	// hello
}
